internal/hcsoci/network.go

```diff
@@ -67,3 +68,89 @@ func GetNamespaceEndpoints(ctx context.Context, netNS string) ([]*hns.HNSEndpoin
 	}
 	return endpoints, nil
 }
+
+// SetupNetworkNamespace adds the network namespace with given nsid into
+// the given UtilityVM and then adds the endpoints associated with that nsid into that
+// namespace.
+func SetupNetworkNamespace(ctx context.Context, hostingSystem *uvm.UtilityVM, nsid string) (err error) {
+	endpoints, err := GetNamespaceEndpoints(ctx, nsid)
+	if err != nil {
+		return err
+	}
+	err = hostingSystem.AddNetNS(ctx, nsid)
+	if err != nil {
+		return err
+	}
+	err = hostingSystem.AddEndpointsToNS(ctx, nsid, endpoints)
+	if err != nil {
+		// Best effort clean up the NS
+		hostingSystem.RemoveNetNS(ctx, nsid)
+		return err
+	}
+	return nil
+}
+
+// Network namespace setup is a bit different for templates and clones.
+// A normal network namespace creation works as follows: When a new Pod is created, HNS is
+// called to create a new namespace and endpoints inside that namespace. Then we hot add
+// this network namespace and then the network endpoints associated with that namespace
+// into this pod UVM. Later when we create containers inside that pod they start running
+// inside this namespace. The information about namespace and endpoints is maintained by
+// the HNS and it is stored in the guest VM's registry (UVM) as well. So if inside the pod
+// we see a namespcae with id 'NSID' then we can query HNS to find all the information
+// about that namespace.
+//
+// When we clone a VM (with containers running inside it) we can hot add a new namespace
+// and endpoints created for that namespace to the uvm but we can't make the existing
+// processes/containers to automatically switch over to this new namespace. To solve this
+// problem when we create a template or a cloned pod we will ask HNS to create a new
+// namespace and endpoints for that UVM but when we actually send a request to hot add
+// that namespace we will change the namespace ID with the default ID that is specifically
+// created for cloning purposes. Similarly, when hot adding an endpoint we will modify
+// this endpoint information to set its network namespace ID to this default ID. This way
+// inside every template and cloned pod the namespace ID will remain same (but each cloned
+// UVM will have a different endpoint) but the HNS will have the actual namespace ID that
+// was created for that UVM. The reason to have a specific default ID is that it will help
+// in debugging namespace related scenarios as it makes it clear that this is a cloned VM
+// and hence has a namespace that was cloned from some template.
```

Comment:
I'm not sure I understand what's going on here. What does changing the namespace ID do for us? do we now have multiple network namespaces in the UVM?

Reply:
Take a look at the updated comment in the new PR and we can discuss more offline if you want.